### PR TITLE
Refactor noxfile

### PR DESCRIPTION
**Pull request by jonparrott**

### CODE REVIEW

**Comment by jonparrott** [comment]
This refactor isolates every samples' run from all others, each sample now has an isolated virtualenv. This will make complete builds slower, but incremental builds shouldn't suffer.

This will also make it easier for us to add user journey tests.

File: nox.py [hunk 1]
```diff
@@ -77,272 +57,223 @@ def collect_sample_dirs(start_dir, blacklist=set(), suffix='_test.py'):
...
+
+def session_missing_tests(session):
+    """Lists all sample directories that do not have tests."""
     session.virtualenv = False
-    requirements = set()
-    requirements_files = list(list_files('.', 'requirements*.txt'))
-    requirements_files.append('./testing/requirements-dev.in')
+    print('The following samples do not have tests:')
+    for sample in set(ALL_SAMPLE_DIRECTORIES) - set(ALL_TESTED_SAMPLES):
+        print('*', sample)
```
Comment (waprin) [inline nox.py]: This is intentional? Expected something like print('*{}').format(sample) .

**Comment by waprin** [comment]
lgtm

File: nox.py [hunk 2]
```diff
@@ -77,272 +57,223 @@ def collect_sample_dirs(start_dir, blacklist=set(), suffix='_test.py'):
...
+def _determine_local_import_names(start_dir):
+    """Determines all import names that should be considered "local".
 
-    gae_root = os.environ.get('GAE_ROOT', tempfile.gettempdir())
-    session.env['GAE_SDK_PATH'] = os.path.join(gae_root, 'google_appengine')
-    session.run('gcprepotools', 'download-appengine-sdk', gae_root)
+    This is used when running the linter to insure that import order is
+    properly checked.
+    """
+    return [
```
Comment (jonparrott) [inline nox.py]: Done

File: nox.py (hunk 1 above)
Comment (jonparrott) [inline nox.py]: Done

**Comment by jonparrott** [comment]
Merging, can address other comments in follow-ups.

File: scripts/travis.sh [hunk 3]
```diff
@@ -6,7 +6,7 @@ if [[ $TRAVIS_SECURE_ENV_VARS == "true" ]]; then
     source ${TRAVIS_BUILD_DIR}/testing/test-env.sh;
     export GOOGLE_APPLICATION_CREDENTIALS=${TRAVIS_BUILD_DIR}/testing/service-account.json
     export GOOGLE_CLIENT_SECRETS=${TRAVIS_BUILD_DIR}/testing/client-secrets.json
-    nox --stop-on-first-error -s lint travis;
+    nox --stop-on-first-error -s lint gae py35;
```
Comment (jerjou) [inline scripts/travis.sh]: Not testing py27 anymore?

Comment (jonparrott) [inline scripts/travis.sh]: We weren't before either.

**Pull request merged**